### ISSUE TITLE
Make usage of _ for private field and prefix for static field optional and not mandatory (as for previous coding conventions)

### ISSUE THREAD

**Issue by PhotoAtomic** [dotnet/docs#25754]
The part related with the camelCase naming convention contains update from this commit 
https://github.com/dotnet/docs/commit/631c596d095e9c0815c838a5b8b8ec79c122dd5b
enforces the following rule
"Use camel casing ("camelCasing") when naming private or internal fields, and prefix them with _."
which in the previous version of the c# coding convention was not stated and instead the private fields in the examples of the previous version of the documentation was always displayed unprefixed

```C#
public class EmployeeRecord
{
    private int employeeId;
    private int department;
    public EmployeeRecord()
    {
    }
    public  EmployeeRecord (int id, int departmentId)
    {
        EmployeeId = id;
        Department = departmentId;
    }
    public int Department
    {
        get {return department;}
        set {department = value;}
    }
    public int EmployeeId
    {
        get {return employeeId;}
        set {employeeId = value;}
    }
    public EmployeeRecord Clone()
    {
        return new EmployeeRecord(employeeId, department);
    }
}
```

this is a quite a breaking change for highly regulated organization that must adhere to the Microsoft's proposed coding convention, because implies the rewrite of a lot of code just for a stylistic rule that wasn't there before, or a tracked deviation from it.

Please make this a **suggestion** and **not mandatory**, as it was in the previous versions of the coding guide lines.

Moreover, previous version directly discourage hungarian notation when naming static fields 
"•	Do not use a Hungarian notation prefix on static field names. "
which is quite in contrast with the new rule
"When working with static fields that are private or internal, use the s_ prefix and for thread static use t_."

We can discuss of what hungarian notation really is (from wikipedia: "Hungarian notation is an identifier naming convention in computer programming, in which the name of a variable or function indicates its **intention** or kind, and in some dialects its type.") , but the spirit of the guideline was "not not use prefix on static fields to indicate intention" the new rule is in the complete opposite direction. 
Also, the same consideration on pre-existing compliant code bases applies: **make these optional, not mandatory** or reconsider it completely.

Moreover, IMHO, _ (underscores) makes code less readable, modern IDE are able to distinguish what is private by the private modifier, _ comes from lower level language where it was necessary to distinguish what to touch and what not.
Fortunately high level OO language such C# don't suffer of these problems and modern IDE are not helped by _ prefix neither (which can be a tip, but absolutely not a recommendation)

By these consideration and the amount of (useless) effort this new rule can implies for certain organization, it is required to make these rules optional and report these as advices using a less mandatory tone in the sentence, or clearly stating that these are suggestions
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 40f42506-f4a5-859a-0266-ded002056d83
* Version Independent ID: f940f2ad-e13d-4d37-6439-f8afcbc4e4c7
* Content: [C# Coding Conventions](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case)
* Content Source: [docs/csharp/fundamentals/coding-style/coding-conventions.md](https://github.com/dotnet/docs/blob/main/docs/csharp/fundamentals/coding-style/coding-conventions.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by Youssef1313** [comment]
It's not really a "new rule". The .NET runtime team has been following this coding convention for years. Also I don't think it's mandatory to follow, feel free to use whatever convention as long as you're consistent. And actually, there can be some coding conventions that differ between different teams in Microsoft. So again, as long as you're consistent you're good to go. This document just reflects what the .NET runtime team uses.

**Comment by PhotoAtomic** [comment]
Hello Youseff, thanks for your reply but I want to describe my situation better so to allow you to better evaluate the impact of this change
* I work for a pharma company: pharma is notoriously an highly regulated environment where everything must be documented (at the point where terms like should, shall, could, can and must are described in a document that inherits from government documents, FDA in this case). Coding convention used to create internal programs must also be documented.
* I'm leading a project started in October last year that is in plan to conclude on December this year.
* the project started by using with the "older" version of the Microsoft coding guidelines for C# where no mention was made to the required usage of the _ for private fields not prefix for static variables.
* the title of the document in object is "C# Coding Conventions" it is meant to be the coding convention for the whole C# community, otherwise the title should be "C# Coding Conventions for the Microsoft Runtime Team" as you said, other teams in Microsoft uses a different coding convention as well, but I'm pretty sure nobody uses PascalCase for privated fields, there was in fact a rule in older naming convention that states that privates fields should not start with a capital letter, should use camelCase, and should not use _ to connect multiple words. As you can see, given that description, starting with _ or not was possible as long as the first letter is in lower case.
* the [new rule](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case) I'm referring to was added the 5 may of the 2021, with the commit linked above, its worth nothing that internally it was used per year. it is in fact a **new** rule from the point of view of a consumer of this document.
* the new rule can be seen as a sort of breaking change because it automatically makes some of the code that was before compliant with the guideline immediately not complaint. In fact it was previously possible to call private member using simple camelCasing, and now by the rule "use camel casing ("camelCasing") when naming private or internal fields, *and prefix them with _.*" it is no more possible because the rule says *and* which is quite mandatory
* similar consideration applies for static, where i'm even more worried because in this case the new rule (applied by the same commit) contraddict in my opion older rule which clearly says "to not use hungarian notation for static". As said we can discuss if this is hungarian notation or not, the problem is that the new rule makes this mandatory "**When** working with static fields that are private or internal, **use the s_ prefix and for thread static use t_.**"
* as sait i'm inthe middle of a project that refer to the Microsoft naming convention as guideline for the coding (there will be a code revision phase where other people will check also that the rules in the guideline are respected) this addition to the official guideline makes the code written so far for this project not compliant.

my proposal (and I can make a PR for this) is to rephrase these rule in a way that makes clear that these new parts are suggestion or example, by using a less mandatory tone.

I would make clear that by the rule introduced in the commit above even the same documentation is inconsistent. Look for at the example at [the reference page for the private keyword](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private) or at the [reference for thread static attribute](https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=net-5.0#examples)

Using this rules as they are many part of the documentation must be updated as well.

I propose to rephrase as follow:

**From:**
"
Use camel casing ("camelCasing") when naming private or internal fields, and prefix them with _.
"
**To:**
"
Use camel casing ("camelCasing") when naming private or internal fields, optionally prefixing them with _.
"

**and:**
"
When working with static fields that are private or internal, use the s_ prefix and for thread static use t_.
"
**To:**
"
When working with static fields that are private or internal, the s_ prefix for static and for t_ prefix for thread static can be used.
"

This way all the teams can continue to use their convention, the new rule is encouraged but who follow the old guideline (like me in my project) have it's code base still compliant.

**Comment by Youssef1313** [comment]
@PhotoAtomic The whole article has a note:

> The guidelines in this article are used by Microsoft to develop samples and documentation. They were adopted from the .NET Runtime, C# Coding Style guidelines. You can use them, or adapt them to your needs. The primary objectives are consistency and readability within your project, team, organization, or company source code.

Is this "general" rule not enough?

The problem with making this specific convention optional is that it's not what the .NET Runtime team do, which is what the docs is following.

@BillWagner @IEvangelist What do you think should be done here?

**Comment by PhotoAtomic** [comment]
Please note that the sentence:
>The guidelines in this article are used by Microsoft to develop **samples and documentation.**

is false already, as I've posted two examples where the documentation is not following these convention.
As said, this is an addition made on 5 may 2021, it was not present before (or I'm unable to find it).



Probably, by reversing the reference between the two docs can make everybody happy.

The [general one](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case) can stay generic enough and the [specific one](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) can detail as much as it wants for the Runtime Team, and this could be possible while the "general one" doesn't dictate rules that are too strict.

The runtime team already has a more detailed and specific document, why not make that inherit from the generic one and detail required additions/deviation? In any case, by already having a specific document for themselves, the Runtime team is safe: I don't think that specific doc need any modification. The problem is only in the general one.

I think we can find a nice way to describe the situation that makes everybody happy :)

**Comment by IEvangelist** [comment]
Hi 👋 @PhotoAtomic

Thank you so much for posting this issue, and providing such great detail. We really appreciate it, and thank you for your time. 

The short response is that this guidance is not going to back step now. I believe that @Youssef1313 said if very well here:

> It's not really a "new rule". The .NET runtime team has been following this coding convention for years. Also I don't think it's mandatory to follow, feel free to use whatever convention as long as you're consistent. And actually, there can be some coding conventions that differ between different teams in Microsoft. So again, as long as you're consistent you're good to go. This document just reflects what the .NET runtime team uses.

This is guidance that reflects what the .NET runtime teams uses. You could use an [ *.editorconfig* file](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/configuration-files), and globally apply changes in a single PR to update all the coding styles appropriately with minimal effort - probably 10 minutes.

**TL;DR;**

This is guidance, you're free to do as you want with it. If an organization is "highly regulated" and adhering to this guidance - they should anticipate occassional updates and plan for them. But again, this is guidance. We've adapted it recently with changes and updates to align with the majority of the source code in our samples, and with the source code in the .NET runtime itself.

> Moreover, IMHO, _ (underscores) makes code less readable, modern IDE are able to distinguish what is private by the private modifier, _ comes from lower level language where it was necessary to distinguish what to touch and what not.

This is your opinion, but my opinion differs. There are two major advantages to using the `_` for these fields:

1. Within a method body, you can type `_` to see all of the fields in the class scope - this is better than typing `this.`, because instead of typing five characters you're only typing one. Also, you'll get the list of fields not other members on the `this` context, such as Methods, events, other non-field members, etc.
2. When assigning a field from a constructor, if the names are the same you can avoid the `this.` - which is again saving you 5x on characters per member:
 
```diff
-private readonly IServiceOne serviceOne;
-private readonly IServiceTwo serviceTwo;
-private readonly IServiceThree serviceThree;

+private readonly IServiceOne _serviceOne;
+private readonly IServiceTwo _serviceTwo;
+private readonly IServiceThree _serviceThree;

public ConsumingService(
    IServiceOne serviceOne,
    IServiceTwo serviceTwo,
    IServiceThree serviceThree)
{
-    this.serviceOne = serviceOne;
-    this.serviceTwo = serviceTwo;
-    this.serviceThree = serviceThree;

+    _serviceOne = serviceOne;
+    _serviceTwo = serviceTwo;
+    _serviceThree = serviceThree;
}
```

It seems as though you're regulated organization is looking for documented coding conventions, if you're choosing to use this document - know that it is a living document and it will evolve. If you want something more static, you are free to create your own document that is static based on these guidelines. To me it sounds like you were trusting these docs for your current project, but now since they have changed, you no longer want to trust them. Again, take them or leave them - I'm sorry but that is your choice.

> The title of the document in object is "C# Coding Conventions" it is meant to be the coding convention for the whole C# community, otherwise the title should be "C# Coding Conventions for the Microsoft Runtime Team".

These are **C# Coding Conventions**, so the title is appropriate and the document calls attention to where these are specifically sourced.

![image](https://user-images.githubusercontent.com/7679720/130460732-3c09ba8a-af8e-4c12-a45f-d2decc7798db.png)

> I think we can find a nice way to describe the situation that makes everybody happy.

I do too, but I believe that we already have. If you're relying on a document that changes for a project that has a deadline - you're going to eventually run into problems like this.

> The new rule can be seen as a sort of breaking change because it automatically makes some of the code that was before compliant with the guideline immediately not complaint. In fact it was previously possible to call private member using simple camelCasing, and now by the rule "use camel casing ("camelCasing") when naming private or internal fields, and prefix them with _." it is no more possible because the rule says and which is quite mandatory.

This is **_not_** mandatory, these are **_not_** rules - they are coding conventions and guidelines. The reality is that we cannot make every developer happy. I'm sorry, but we're going to close this issue as the document is currently correct.

**Comment by PhotoAtomic** [comment]
Hello IEvangelist, 

By your point of view you should now go and fix all the documentation in the repository that makes example using a convention different from the one dictated in the page in object, right?

Plus, fix also VisualStudio auto generation capability which is more than happy to generate private fields with the same name of the parameter in the constructor and then assign values using "this".
Here below is in fact a piece of code generated by VS2019 automatically using the refactoring function "generate class in a new file" 
```C#
namespace ConsoleApp1
{
    internal class MyClass
    {
        private string myParameter;

        public MyClass(string myParameter)
        {
            this.myParameter = myParameter;
        }
    }
}
```

for the code below
```C#
var MyParameter = "value";
var myInstance = new MyClass(MyParameter);
```

I'm sure it is not the only tool that does it. If I follow your reasoning I can use the code generation function to save keystrokes  and then manually fix the "this" keyword and rename the private fields with the "_"prefix: It is not a good deal and I think your reasoning here is quite convoluted.

You keep repeating that these are not rules and they are not mandatory, but you refuse to make them more general (no sorry, report them as they were before may 5 2021) when the team of interest (Runtime team) already have a more specific document.
Why?

The fact that is a living document doesn't means you can change it when you want without side effect. If this were code we would have had a breaking change here, I'm pretty sure you know what it is.

Last, for earlier projects I already have a copy of the guidelines in effect at that time, and that was where I was able to report which the difference are and what the text was used to be. This time I've notice the problem because the "update" occurred in the middle of the project (maybe thanks to the big "s***w you" attached to it...)

I'll then wait for the update on all the example in the documentation, good luck (or maybe you'll remove the sentence at the top where is reported that those guidelines are used to write the examples in the doc... who knows?).

Again the big box at the top reports 
>The primary objectives are consistency and readability within your project, team, organization, or company source code.

Tell me: how can a company can reach consistency by adopting a document that randomly adds more restrictive rules? I'm going to make a copy of this document ad adapt it where needed, thanks for the advice I've never thought about it.

**Comment by IEvangelist** [comment]
Hi again @PhotoAtomic, happy to help, take care.
